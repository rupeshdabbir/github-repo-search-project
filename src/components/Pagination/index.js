// @flow

import React from 'react';
import {Link} from 'react-router-dom';
import first from 'lodash/fp/first';
import flow from 'lodash/fp/flow';
import {
  StyleSheet,
  css,
} from 'aphrodite/no-important';
import classNames from 'classnames';
import 'suitcss-utils-flex/lib/flex.css';
import NextIcon from './arrow-right.svg';
import PrevIcon from './arrow-left.svg';

type Props = {
  next: Object,
  prev: Object,
};

const getSearchQuery = flow(
  url => url.match(/(\?.+)/),
  first
);

function renderLink({url, extraClassName, rel, text, icon: Icon}: Object) {
  if (!url) {return null;}
  const search = getSearchQuery(url);

  const linkClass = classNames(
    css(styles.Pagination_link),
    'u-flex',
    'u-flexAlignItemsCenter',
    extraClassName
  );

  return (
    <Link
      to={{search}}
      className={linkClass}
      rel={rel}
    >
      <span>{text}</span>
      <Icon width={40} height={40} />
    </Link>
  );
}

function Pagination({prev, next}: Props) {
  const prevLink = renderLink({
    url: prev.url,
    extraClassName: 'u-flexRowReverse u-flexExpandRight',
    rel: 'prev',
    text: 'Previous',
    icon: PrevIcon,
  });

  const nextLink = renderLink({
    url: next.url,
    extraClassName: 'u-flexExpandLeft',
    rel: 'next',
    text: 'Next',
    icon: NextIcon,
  });

  return (
    <div className="u-flex u-flexJustifyCenter">
      {prevLink}
      {nextLink}
    </div>
  );
}

Pagination.defaultProps = {
  next: {},
  prev: {},
};

const styles = StyleSheet.create({
  Pagination_link: {
    textDecoration: 'none',
    color: 'inherit',
  },
});

export default Pagination;
