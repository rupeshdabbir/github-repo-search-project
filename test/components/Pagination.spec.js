import React from 'react';
import renderer from 'react-test-renderer';
import {
  StaticRouter as Router,
} from 'react-router-dom';
import {StyleSheetTestUtils} from 'aphrodite';

import Pagination from 'components/Pagination';

function createComponent(props) {
  return renderer.create(
    <Router context={{}}>
      <Pagination {...props} />
    </Router>
  );
}

describe('Component: Pagination', () => {

  beforeAll(() => {
    StyleSheetTestUtils.suppressStyleInjection();
  });

  afterAll(() => {
    StyleSheetTestUtils.clearBufferAndResumeStyleInjection();
  });

  describe('when passed a next prop', () => {
    it('should render a next link', () => {
      const props = {
        next: {
          url: 'https://api.github.com/search/users?q=simon&page=2',
        },
      };
      const component = createComponent(props);
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe('when passed a prev prop', () => {
    it('should render a previous link', () => {
      const props = {
        prev: {
          url: 'https://api.github.com/search/users?q=simon&page=1',
        },
      };
      const component = createComponent(props);
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe('when passed a prev and next prop', () => {
    it('should render a previous and next link', () => {
      const props = {
        prev: {
          url: 'https://api.github.com/search/users?q=simon&page=1',
        },
        next: {
          url: 'https://api.github.com/search/users?q=simon&page=2',
        },
      };
      const component = createComponent(props);
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

});
