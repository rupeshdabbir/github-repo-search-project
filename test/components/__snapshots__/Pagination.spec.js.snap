// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Pagination when passed a next prop should render a next link 1`] = `
<div
  className="u-flex u-flexJustifyCenter"
>
  <a
    className="Pagination_link_18deiwy u-flex u-flexAlignItemsCenter u-flexExpandLeft"
    href="/?q=simon&page=2"
    onClick={[Function]}
    rel="next"
  >
    <span>
      Next
    </span>
    <svg />
  </a>
</div>
`;

exports[`Component: Pagination when passed a prev and next prop should render a previous and next link 1`] = `
<div
  className="u-flex u-flexJustifyCenter"
>
  <a
    className="Pagination_link_18deiwy u-flex u-flexAlignItemsCenter u-flexRowReverse u-flexExpandRight"
    href="/?q=simon&page=1"
    onClick={[Function]}
    rel="prev"
  >
    <span>
      Previous
    </span>
    <svg />
  </a>
  <a
    className="Pagination_link_18deiwy u-flex u-flexAlignItemsCenter u-flexExpandLeft"
    href="/?q=simon&page=2"
    onClick={[Function]}
    rel="next"
  >
    <span>
      Next
    </span>
    <svg />
  </a>
</div>
`;

exports[`Component: Pagination when passed a prev prop should render a previous link 1`] = `
<div
  className="u-flex u-flexJustifyCenter"
>
  <a
    className="Pagination_link_18deiwy u-flex u-flexAlignItemsCenter u-flexRowReverse u-flexExpandRight"
    href="/?q=simon&page=1"
    onClick={[Function]}
    rel="prev"
  >
    <span>
      Previous
    </span>
    <svg />
  </a>
</div>
`;
